[HUDI-7176] Add file group reader test framework

linliu-code opened the pull request:
Our goal is to add a framework for unit tests at the level of file group reading. That is,
we don't need to involve any engine specific logic during the read, such that we could

1. Simplify the test process by providing unified and simple APIs,
2. Impose more granular read process for testing, and
3. Improve the test performance since no engine specific logic are involved.

We will add the framework step by step.

### Change Logs

Steps:
1. Add a builder class to construct file group reader.

### Impact

1. Simplify the process to construct file group reader level tests;
5. Improve the test performance by removing engine specific logic.

### Risk level (write none, low medium or high below)

None.

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

linliu-code commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/reader/DataGenerationPlan.java:
```diff
@@ -0,0 +1,122 @@
+ */
+public class DataGenerationPlan {
+  // The values for "_row_key" field.
+  private final List<String> recordKeys;
```
If needed in the future, we can add that functionality.

linliu-code commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/reader/HoodieFileSliceTestUtils.java:
```diff
@@ -0,0 +1,437 @@
+
+public class HoodieFileSliceTestUtils {
+  public static final String FORWARD_SLASH = "/";
+  public static final String PERIOD = ".";
```
Do you mean why var for these literals?

linliu-code replied:
Just a style. Here we can either use the literals directly or use variables. :)

linliu-code commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/reader/HoodieFileSliceTestUtils.java:
```diff
@@ -0,0 +1,437 @@
+        hadoopConf,
+        0.1,
+        true);
+
```
Added a task to track this problem: [HUDI-7214](https://issues.apache.org/jira/browse/HUDI-7214)

yihua commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/reader/DataGenerationPlan.java:
```diff
@@ -0,0 +1,126 @@
+ * 1. One plan generates one file, either a base file, or a log file.
+ * 2. One file contains one operation, e.g., insert, delete, or update.
+ */
+public class DataGenerationPlan {
```
Is this going to be integrated with `HoodieTestTable` or similar to generate test tables without using transactional writes?

linliu-code replied:
So far I haven't seen any necessity for it. But in the future if we need, we can extend it.

yihua replied:
Let's have only one test framework to generate test tables.  Let's do that in the follow-up PR.  Also add a guideline on how to use this new test framework to construct tests that needs preparation of a Hudi table with data files.

yihua commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/reader/HoodieFileSliceTestUtils.java:
```diff
@@ -0,0 +1,440 @@
+import static org.apache.hudi.common.testutils.reader.DataGenerationPlan.OperationType.DELETE;
+import static org.apache.hudi.common.testutils.reader.DataGenerationPlan.OperationType.INSERT;
+
+public class HoodieFileSliceTestUtils {
```
Should this be part of `HoodieTestTable` instead of utils class/methods so one can easily write a Hudi table on storage for testing?

linliu-code replied:
I plan to have another PR to further extract the framework, I can integrate it to the `HoodieTestTable` then.

linliu-code commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/reader/HoodieTestReaderContext.java:
```diff
@@ -0,0 +1,163 @@
+    return builder.build();
+  }
+
+  @Override
+  public ClosableIterator<IndexedRecord> mergeBootstrapReaders(ClosableIterator<IndexedRecord> skeletonFileIterator, ClosableIterator<IndexedRecord> dataFileIterator) {
+    return null;
+  }
+
+  @Override
+  public UnaryOperator<IndexedRecord> projectRecord(Schema from, Schema to) {
+    return null;
+  }
```
Yeah, this is introduced in Jon recently. I have implemented in another PR.

linliu-code commented:
@yihua, I have addressed your comments. My overall plan is to land this PR, and use another PR to further simplify this framework and integrate more with other existing modules. How does it sound?

yihua reviewed (approved):
+1
